From 0834f4f350bcda02dd09005be32ee9d32987453a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Lasse=20Brudeskar=20Vik=C3=A5s?= <bvx89.at.cs@gmail.com>
Date: Sun, 19 Jan 2014 22:01:57 +0100
Subject: [PATCH 1/2] Add Download Speeds in notification

This commit adds the ability to show the current total download speed

Screenshots: http://goo.gl/M3eRNR

ps7 : remove download site, and make speed to bit/sec
ps8 : make speed to byte/sec and cleanup
ps9 : revert to B/sec
ps10: fix some formatting issues
ps13: rebased

Signed-off-by: Chet Kener <Cl3Kener@gmail.com>

Conflicts:
	res/values/cm_strings.xml
	src/com/android/providers/downloads/DownloadNotifier.java

Change-Id: I801dbe61c7ee59d0c1d14d5851ad6dc3a7678499
Signed-off-by: Chet Kener <Cl3Kener@gmail.com>
---
 res/values/cm_plurals.xml                          | 33 ++++++++++++++++++++++
 res/values/cm_strings.xml                          | 19 +++++++++++++
 .../providers/downloads/DownloadNotifier.java      | 32 +++++++++++++++++----
 3 files changed, 78 insertions(+), 6 deletions(-)
 create mode 100644 res/values/cm_plurals.xml
 create mode 100644 res/values/cm_strings.xml

diff --git a/res/values/cm_plurals.xml b/res/values/cm_plurals.xml
new file mode 100644
index 0000000..b201ff5
--- /dev/null
+++ b/res/values/cm_plurals.xml
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     Copyright (C) 2014 The CyanogenMod Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
+    <!-- Phrase describing a time duration using seconds [CHAR LIMIT=16] -->
+    <plurals name="duration_seconds">
+        <item quantity="one"><xliff:g id="count">%d</xliff:g> sec</item>
+        <item quantity="other"><xliff:g id="count">%d</xliff:g> secs</item>
+    </plurals>
+    <!-- Phrase describing a time duration using minutes [CHAR LIMIT=16] -->
+    <plurals name="duration_minutes">
+        <item quantity="one"><xliff:g id="count">%d</xliff:g> min</item>
+        <item quantity="other"><xliff:g id="count">%d</xliff:g> mins</item>
+    </plurals>
+    <!-- Phrase describing a time duration using hours [CHAR LIMIT=16] -->
+    <plurals name="duration_hours">
+        <item quantity="one"><xliff:g id="count">%d</xliff:g> hr</item>
+        <item quantity="other"><xliff:g id="count">%d</xliff:g> hrs</item>
+    </plurals>
+</resources>
diff --git a/res/values/cm_strings.xml b/res/values/cm_strings.xml
new file mode 100644
index 0000000..fb6f06c
--- /dev/null
+++ b/res/values/cm_strings.xml
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+     Copyright (C) 2013-2014 The CyanogenMod Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
+    <string name="download_speed_text"><xliff:g id="text">%1$s</xliff:g>, <xliff:g id="size" example="230 kB">%2$s</xliff:g>/s</string>
+</resources>
diff --git a/src/com/android/providers/downloads/DownloadNotifier.java b/src/com/android/providers/downloads/DownloadNotifier.java
index 60c249f..3af9746 100644
--- a/src/com/android/providers/downloads/DownloadNotifier.java
+++ b/src/com/android/providers/downloads/DownloadNotifier.java
@@ -35,6 +35,7 @@ import android.os.SystemClock;
 import android.provider.Downloads;
 import android.text.TextUtils;
 import android.text.format.DateUtils;
+import android.text.format.Formatter;
 import android.util.Log;
 import android.util.LongSparseLongArray;
 
@@ -205,6 +206,7 @@ public class DownloadNotifier {
             // Calculate and show progress
             String remainingText = null;
             String percentText = null;
+            String speedAsSizeText = null;
             if (type == TYPE_ACTIVE) {
                 long current = 0;
                 long total = 0;
@@ -225,8 +227,26 @@ public class DownloadNotifier {
 
                     if (speed > 0) {
                         final long remainingMillis = ((total - current) * 1000) / speed;
+                        final int duration, durationResId;
+
+                        // This duplicates DateUtils.formatDuration(), but uses our
+                        // abbreviated plurals.
+                        if (remainingMillis >= DateUtils.HOUR_IN_MILLIS) {
+                            duration = (int) ((remainingMillis + 1800000)
+                                    / DateUtils.HOUR_IN_MILLIS);
+                            durationResId = R.plurals.duration_hours;
+                        } else if (remainingMillis >= DateUtils.MINUTE_IN_MILLIS) {
+                            duration = (int) ((remainingMillis + 30000)
+                                    / DateUtils.MINUTE_IN_MILLIS);
+                            durationResId = R.plurals.duration_minutes;
+                        } else {
+                            duration = (int) ((remainingMillis + 500)
+                                    / DateUtils.SECOND_IN_MILLIS);
+                            durationResId = R.plurals.duration_seconds;
+                        }
                         remainingText = res.getString(R.string.download_remaining,
-                                DateUtils.formatDuration(remainingMillis));
+                                res.getQuantityString(durationResId, duration, duration));
+                        speedAsSizeText = Formatter.formatFileSize(mContext, speed);
                     }
 
                     final int percent = (int) ((current * 100) / total);
@@ -244,10 +264,9 @@ public class DownloadNotifier {
                 builder.setContentTitle(getDownloadTitle(res, info));
 
                 if (type == TYPE_ACTIVE) {
-                    if (!TextUtils.isEmpty(info.mDescription)) {
-                        builder.setContentText(info.mDescription);
-                    } else {
-                        builder.setContentText(remainingText);
+                    if (speedAsSizeText != null) {
+                        builder.setContentText(res.getString(R.string.download_speed_text,
+                                remainingText, speedAsSizeText));
                     }
                     builder.setContentInfo(percentText);
 
@@ -277,7 +296,8 @@ public class DownloadNotifier {
                     builder.setContentTitle(res.getQuantityString(
                             R.plurals.notif_summary_active, cluster.size(), cluster.size()));
                     builder.setContentText(remainingText);
-                    builder.setContentInfo(percentText);
+                    builder.setContentInfo(res.getString(R.string.download_speed_text,
+                                percentText, speedAsSizeText));
                     inboxStyle.setSummaryText(remainingText);
 
                 } else if (type == TYPE_WAITING) {
-- 
2.5.0

